### PR TITLE
SAAS-2290: Plugin to check AWS ElasticSearch Cluster status

### PR DESCRIPTION
**Pull request by UmairAjmal**

### CODE REVIEW

File: collectors/aws/collector.js [hunk 1]
```diff
@@ -853,6 +853,11 @@ var postcalls = [
                 filterValue: 'TrailARN'
             }
         },
+        CloudWatch: {
+            getEsMetricStatistics: {
+                override: true,
```
Comment (AkhtarAmir) [inline collectors/aws/collector.js:858]: add relies on service

File: collectors/aws/cloudwatch/getEsMetricStatistics.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        collection.cloudwatch.getEsMetricStatistics[AWSConfig.region][domain.DomainName] = {};
+        var endTime = new Date();
+        var startTime = new Date();
+        startTime.setTime(startTime.getTime() - 1000);
+        var params = {
+            'MetricName': 'ClusterStatus.Red',
+            'Namespace':'AWS/ES',
+            'StartTime': startTime.toISOString(),
+            'EndTime': endTime.toISOString(),
+            'Period': 60,
```
Comment (AkhtarAmir) [inline collectors/aws/cloudwatch/getEsMetricStatistics.js:17]: Update these

File: plugins/aws/es/esClusterStatus.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+            }
+
+            if (!listDomainNames.data.length){
+                helpers.addResult(results, 0, 'No ES domains found', region);
+                return rcb();
+            }
+
+            var found = false;
+
+            async.each(listDomainNames.data, function(domain, dcb){
```
Comment (AkhtarAmir) [inline plugins/aws/es/esClusterStatus.js:37]: Add domain.DomainName check

File: plugins/aws/es/esClusterStatus.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+               
+                if (!getMetricStats ||
+                    getMetricStats.err ||
+                    !getMetricStats.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for ES domain config: ${helpers.addError(getMetricStats)}`, region);
+                    return dcb();
+                }
+
+                if (!getMetricStats.data.Datapoints.length) return dcb();
```
Comment (AkhtarAmir) [inline plugins/aws/es/esClusterStatus.js:49]: Remove this

File: plugins/aws/es/esClusterStatus.js [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                }
+                
+                dcb();
+            }, function(){
+                rcb();
+            });
+
+            if (!found){
+                helpers.addResult(results, 0, 'All ES Clusters are healthy', region);
+            }
```
Comment (AkhtarAmir) [inline plugins/aws/es/esClusterStatus.js:68]: Remove this

File: plugins/aws/es/esClusterStatus.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+                    return dcb();
+                }
+
+                if (!getMetricStats.data.Datapoints.length) return dcb();
+                var maximumValue = 0;
+                getMetricStats.data.Datapoints.forEach((dataPoint) => {
+                    if (maximumValue < dataPoint.Maximum) maximumValue = dataPoint.Maximum;
+                });
+
+                if (maximumValue > 1){
```
Comment (AkhtarAmir) [inline plugins/aws/es/esClusterStatus.js:55]: Add else case

File: plugins/aws/es/esClusterStatus.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+            async.each(listDomainNames.data, function(domain, dcb){
+                var getMetricStats = helpers.addSource(cache, source,
+                    ['cloudwatch', 'getEsMetricStatistics', region, domain.DomainName]);
+               
+                if (!getMetricStats ||
+                    getMetricStats.err ||
+                    !getMetricStats.data) {
+                    helpers.addResult(results, 3,
+                        `Unable to query for ES domain config: ${helpers.addError(getMetricStats)}`, region);
```
Comment (AkhtarAmir) [inline plugins/aws/es/esClusterStatus.js:45]:
```suggestion
                        `Unable to query for ES domain metric stats: ${helpers.addError(getMetricStats)}`, region);
```

**Review by AkhtarAmir — CHANGES_REQUESTED** [review]

**Review by AkhtarAmir — APPROVED** [review]

**Review by AkhtarAmir — APPROVED** [review]

File: collectors/aws/cloudwatch/getEsMetricStatistics.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
+var AWS = require('aws-sdk');
+var async = require('async');
+
+module.exports = function(AWSConfig, collection, callback) {
+    var cloudwatch = new AWS.CloudWatch(AWSConfig);
+   
+    async.forEach(collection.es.listDomainNames[AWSConfig.region].data, function(domain, cb){
```
Comment (giorod3) [inline collectors/aws/cloudwatch/getEsMetricStatistics.js:7]:
```suggestion
    async.eachLimit(collection.es.listDomainNames[AWSConfig.region].data, 10,  function(domain, cb){        
```

**Review by giorod3 — CHANGES_REQUESTED** [review]
just the one async limit change

**Review by giorod3 — APPROVED** [review]

**Pull request merged**